client/src/main/scala/org/apache/celeborn/client/ChangePartitionManager.scala

```diff
@@ -79,14 +81,18 @@ class ChangePartitionManager(
                 batchHandleChangePartitionExecutors.submit {
                   new Runnable {
                     override def run(): Unit = {
-                      val distinctPartitions = requests.synchronized {
-                        // For each partition only need handle one request
-                        requests.asScala.filter { case (partitionId, _) =>
-                          !inBatchPartitions.get(shuffleId).contains(partitionId)
-                        }.map { case (partitionId, request) =>
-                          inBatchPartitions.get(shuffleId).add(partitionId)
-                          request.asScala.toArray.maxBy(_.epoch)
-                        }.toArray
+                      val distinctPartitions = {
+                        val requestSet = inBatchPartitions.get(shuffleId)
+                        requests.asScala.map { case (partitionId, request) =>
+                          locks(partitionId % locks.length).synchronized {
+                            if (!inBatchPartitions.contains(partitionId)) {
```

Comment:
Do you mean `requestSet.contains(partitionId)` here?

Since `requestSet` is just `util.HashSet`, it's not thread safe. Multiple threads can concurrently modify`requestSet` if they are processing different partition ids, which I think may cause undefined behavior.

Reply:
yeah, I meant requestSet